### PR TITLE
Notification on announcement date

### PR DESCRIPTION
**Pull request by hedgehog1833**
Finally the last PR for all the announcement stuff. It's already quite old so there are lots of commits here, sorry.
The NotificationService now has somewhat duplicate code, it's not much but if you have a cool idea how to fix that I'd be open for suggestions.

### CODE REVIEW

**Comment by hedgehog1833** [comment]
[Make Notifications configurable](https://trello.com/c/hlDkMGvu/255-make-notifications-configurable)

File: webapp/src/main/java/rocks/metaldetector/service/email/TodaysAnnouncementsEmail.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package rocks.metaldetector.service.email;
+
+import rocks.metaldetector.butler.facade.dto.ReleaseDto;
+import rocks.metaldetector.config.constants.ViewNames;
+
+import java.util.List;
+
+public final class TodaysAnnouncementsEmail extends AbstractEmail {
+
+  public static final String SUBJECT = "Your announcements";
```
Comment (DanielW1987) [inline webapp/src/main/java/rocks/metaldetector/service/email/TodaysAnnouncementsEmail.java:10]: Maybe something like "New release announcements"

File: webapp/src/main/java/rocks/metaldetector/service/notification/NotificationServiceImpl.java [hunk 2]
```diff
@@ -62,7 +63,23 @@ public void notifyOnReleaseDate() {
...
+  @Transactional(readOnly = true)
+  public void notifyOnAnnouncementDate() {
+    var now = LocalDate.now();
+    List<ReleaseDto> todaysAnnouncedReleases = releaseService.findAllReleases(Collections.emptyList(), new TimeRange(now, null)).stream()
+        .filter(release -> release.getAnnouncementDate().isEqual(now))
+        .collect(Collectors.toList());
+
+    notificationConfigRepository.findAll().stream()
+        .filter(config -> config.getUser().isEnabled() &&
+                          config.getNotify() &&
```
Comment (DanielW1987) [inline webapp/src/main/java/rocks/metaldetector/service/notification/NotificationServiceImpl.java:80]: We don't need the `config.getNotify() &&` anymore. Please check also `notifyOnReleaseDate`.

File: webapp/src/main/java/rocks/metaldetector/service/notification/NotificationServiceImpl.java [hunk 3]
```diff
@@ -128,6 +144,20 @@ private void releaseDateNotification(NotificationConfigEntity notificationConfig
     }
   }
 
+  private void announcementDateNotification(NotificationConfigEntity notificationConfigEntity, List<ReleaseDto> releases) {
```
Comment (DanielW1987) [inline webapp/src/main/java/rocks/metaldetector/service/notification/NotificationServiceImpl.java:147]: This is  very redundant here.

You should write a method that accepts a BiFunction to create the email:

```java
doNotifyOnSpecificDate(notificationConfigEntity, List<ReleaseDto> releases, (UserEntity user, List<ReleaseDto> filteredReleases) -> {
    new TodaysAnnouncementsEmail(user.getEmail(), user.getUsername(), filteredReleases)
});
```

File: webapp/src/main/resources/templates/email/todays-announcements-email.html [hunk 4]
```diff
@@ -0,0 +1,36 @@
+<!DOCTYPE html>
+<html layout:decorate="~{email/fragments/layout}"
+      xmlns:layout="http://www.ultraq.net.nz/thymeleaf/layout"
+      xmlns:th="http://www.w3.org/1999/xhtml"
+      lang="en">
+<head>
+    <title>Today's metal album announcements</title>
```
Comment (DanielW1987) [inline webapp/src/main/resources/templates/email/todays-announcements-email.html:7]: This as a subject is also good.

**Review by DanielW1987 — COMMENTED** [review]
Please use BiFunction to reduce the code duplication

File: webapp/src/main/java/rocks/metaldetector/service/notification/NotificationServiceImpl.java (hunk 3 above)
Comment (hedgehog1833) [reply webapp/src/main/java/rocks/metaldetector/service/notification/NotificationServiceImpl.java:147]: Nice, that's what I was looking for :)

**Review by DanielW1987 — APPROVED** [review]

**Pull request merged**